File: node/src/drivers/network/tcp/tcp.go
package tcp

import (
	"crypto/tls"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"kasper/src/abstract/adapters/signaler"
	"kasper/src/abstract/models/core"
	"kasper/src/abstract/models/trx"
	"kasper/src/shell/utils/crypto"
	"kasper/src/shell/utils/future"
	"log"
	"net"
	"strings"
	"sync"

	iaction "kasper/src/abstract/models/action"

	packetmodel "kasper/src/abstract/models/packet"

	cmap "github.com/orcaman/concurrent-map/v2"
)

type Socket struct {
	Id     string
	Lock   sync.Mutex
	Conn   net.Conn
	Buffer [][]byte
	Ack    bool
	app    core.ICore
}

type Tcp struct {
	app     core.ICore
	sockets *cmap.ConcurrentMap[string, *Socket]
}

func (t *Tcp) Listen(port int, tlsConfig *tls.Config) {
	future.Async(func() {
		ln, err := tls.Listen("tcp", fmt.Sprintf(":%d", port), tlsConfig)
		if err != nil {
			log.Fatalf("failed to listen: %v", err)
		}
		defer ln.Close()
		log.Println("Clients' TLS server listening on port ", port)

		for {
			conn, err := ln.Accept()
			if err != nil {
				fmt.Println(err)
				continue
			}
			log.Println("new client connected")
			future.Async(func() { t.handleConnection(conn) }, false)
		}
	}, true)
}

func (t *Tcp) listenForPackets(socket *Socket) {
	defer func() {
		t.sockets.Remove(socket.Id)
		socket.Conn.Close()
	}()
	origin := strings.Split(socket.Conn.RemoteAddr().String(), ":")[0]
	lenBuf := make([]byte, 4)
	buf := make([]byte, 1024)
	nextBuf := make([]byte, 2048)
	readCount := 0
	oldReadCount := 0
	enough := false
	beginning := true
	length := 0
	readLength := 0
	remainedReadLength := 0
	var readData []byte
	for {
		if !enough {
			var err error
			readLength, err = socket.Conn.Read(buf)
			if err != nil {
				log.Println(origin, err)
				log.Println(origin, "socket had error and closed")
				return
			}
			func() {
				socket.Lock.Lock()
				defer socket.Lock.Unlock()
				log.Println(origin, "stat 0: reading data...")

				log.Println(origin, buf[0:readLength])

				readCount += readLength
				copy(nextBuf[remainedReadLength:remainedReadLength+readLength], buf[0:readLength])
				remainedReadLength += readLength

				log.Println(origin, nextBuf[0:readLength])

				log.Println(origin, "stat 1:", readLength, oldReadCount, readCount, remainedReadLength)
			}()
		}

		if beginning {
			if readCount >= 4 {
				func() {
					socket.Lock.Lock()
					defer socket.Lock.Unlock()
					log.Println(origin, "stating stat 2...")
					copy(lenBuf, nextBuf[0:4])
					log.Println(origin, "nextBuf", nextBuf[0:4])
					log.Println(origin, "lenBuf", lenBuf[0:4])
					remainedReadLength -= 4
					copy(nextBuf[0:remainedReadLength], nextBuf[4:remainedReadLength+4])
					length = int(binary.BigEndian.Uint32(lenBuf))
					if length > 20000000 {
						return
					}
					readData = make([]byte, length)
					readCount -= 4
					beginning = false
					enough = true

					log.Println(origin, "stat 2:", remainedReadLength, length, readCount)
				}()

			} else {
				enough = false
			}
		} else {
			if remainedReadLength == 0 {
				enough = false
			} else if readCount >= length {
				func() {
					socket.Lock.Lock()
					defer socket.Lock.Unlock()
					log.Println(origin, "stating stat 3...")
					log.Println(origin, "stat 3 step 1", oldReadCount, length)
					copy(readData[oldReadCount:length], nextBuf[0:length-oldReadCount])
					log.Println(origin, "stat 3 step 2", readLength, readCount, length)
					readCount -= length
					copy(nextBuf[0:readCount], nextBuf[length-oldReadCount:(length-oldReadCount)+readCount])
					log.Println(origin, "nextBuf", nextBuf[0:readCount])
					log.Println(origin, "stat 3 step 3", readCount, length)
					remainedReadLength = readCount
					log.Println(origin, "packet received")
					packet := make([]byte, length)
					copy(packet, readData)
					log.Println(origin, "stat 3 step 4")
					oldReadCount = 0
					enough = true
					beginning = true

					log.Println(origin, "stat 3:", remainedReadLength, oldReadCount, readCount)

					future.Async(func() {
						socket.processPacket(packet)
					}, false)
				}()
			} else {
				func() {
					socket.Lock.Lock()
					defer socket.Lock.Unlock()
					log.Println(origin, "stating stat 4...")

					copy(readData[oldReadCount:oldReadCount+(readCount-oldReadCount)], nextBuf[0:readCount-oldReadCount])
					remainedReadLength = 0
					oldReadCount = readCount
					enough = true

					log.Println(origin, "stat 4:", remainedReadLength)
				}()
			}
		}
	}
}

func (t *Tcp) handleConnection(conn net.Conn) {
	socket := &Socket{Id: crypto.SecureUniqueString(), Buffer: [][]byte{}, Conn: conn, app: t.app, Ack: true}
	t.sockets.Set(strings.Split(conn.RemoteAddr().String(), ":")[0], socket)
	future.Async(func() {
		t.listenForPackets(socket)
	}, false)
}

func (t *Socket) writeUpdate(key string, updatePack any, writeRaw bool) {

	log.Println("preparing update...")

	keyBytes := []byte(key)
	keyBytesLen := make([]byte, 4)
	binary.BigEndian.PutUint32(keyBytesLen, uint32(len(keyBytes)))

	var b3 []byte
	if writeRaw {
		b3 = updatePack.([]byte)
	} else {
		var err error
		b3, err = json.Marshal(updatePack)
		if err != nil {
			log.Println(err)
			return
		}
	}

	packet := make([]byte, 1+len(keyBytesLen)+len(keyBytes)+len(b3))
	pointer := 1

	packet[0] = 0x01

	copy(packet[pointer:pointer+len(keyBytesLen)], keyBytesLen[:])
	pointer += len(keyBytesLen)
	copy(packet[pointer:pointer+len(keyBytes)], keyBytes[:])
	pointer += len(keyBytes)

	copy(packet[pointer:pointer+len(b3)], b3[:])
	pointer += len(b3)

	log.Println("appending to buffer...")

	t.Lock.Lock()
	defer t.Lock.Unlock()

	t.Buffer = append(t.Buffer, packet)
	t.pushBuffer()
}

func (t *Socket) writeResponse(requestId string, resCode int, response any, writeRaw bool) {

	log.Println("preparing response...")

	b1 := []byte(requestId)
	b1Len := make([]byte, 4)
	binary.BigEndian.PutUint32(b1Len, uint32(len(b1)))

	b2 := make([]byte, 4)
	binary.BigEndian.PutUint32(b2, uint32(resCode))

	var b3 []byte
	if writeRaw {
		b3 = response.([]byte)
	} else {
		var err error
		b3, err = json.Marshal(response)
		if err != nil {
			log.Println(err)
			return
		}
	}

	packet := make([]byte, 1+len(b1Len)+len(b1)+len(b2)+len(b3))
	pointer := 1

	packet[0] = 0x02

	copy(packet[pointer:pointer+len(b1Len)], b1Len[:])
	pointer += len(b1Len)
	copy(packet[pointer:pointer+len(b1)], b1[:])
	pointer += len(b1)

	copy(packet[pointer:pointer+len(b2)], b2[:])
	pointer += len(b2)

	copy(packet[pointer:pointer+len(b3)], b3[:])
	pointer += len(b3)

	log.Println("appending to buffer...")

	t.Lock.Lock()
	defer t.Lock.Unlock()

	t.Buffer = append(t.Buffer, packet)
	t.pushBuffer()
}

func (t *Socket) pushBuffer() {
	log.Println("pushing buffer to client...", t.Ack, len(t.Buffer))
	if t.Ack {
		if len(t.Buffer) > 0 {
			t.Ack = false
			packetLen := make([]byte, 4)
			binary.BigEndian.PutUint32(packetLen, uint32(len(t.Buffer[0])))
			_, err := t.Conn.Write(packetLen)
			if err != nil {
				t.Ack = true
				log.Println(err)
				return
			}
			_, err = t.Conn.Write(t.Buffer[0])
			if err != nil {
				t.Ack = true
				log.Println(err)
				return
			}
		}
	}
}

func (t *Socket) connectListener(uid string) *signaler.Listener {
	t.app.Tools().Signaler().Lock()
	defer t.app.Tools().Signaler().Unlock()
	lis, found := t.app.Tools().Signaler().Listeners().Get(uid)
	if !found {
		lis = &signaler.Listener{
			Id:      uid,
			Paused:  false,
			DisTime: 0,
			Signal: func(key string, b any) {
				if b != nil {
					t.writeUpdate(key, b, true)
				}
			},
		}
	}
	t.Ack = true
	return lis
}

func (t *Socket) processPacket(packet []byte) {
	if len(packet) == 1 && packet[0] == 0x01 {
		send := func() {
			t.Lock.Lock()
			defer t.Lock.Unlock()
			t.Ack = true
			if len(t.Buffer) > 0 {
				t.Buffer = t.Buffer[1:]
				t.pushBuffer()
			}
		}
		send()
		return
	}
	pointer := 0
	signatureLength := int(binary.BigEndian.Uint32(packet[pointer : pointer+4]))
	log.Println("signature length:", signatureLength)
	if signatureLength > 20000000 {
		return
	}
	pointer += 4
	signature := string(packet[pointer : pointer+signatureLength])
	pointer += signatureLength
	log.Println("signature:", signature)
	userIdLength := int(binary.BigEndian.Uint32(packet[pointer : pointer+4]))
	pointer += 4
	log.Println("userId length:", userIdLength)
	if userIdLength > 20000000 {
		return
	}
	userId := string(packet[pointer : pointer+userIdLength])
	pointer += userIdLength
	log.Println("userId:", userId)
	pathLength := int(binary.BigEndian.Uint32(packet[pointer : pointer+4]))
	pointer += 4
	log.Println("path length:", pathLength)
	if pathLength > 20000000 {
		return
	}
	path := string(packet[pointer : pointer+pathLength])
	pointer += pathLength
	log.Println("path:", path)
	packetIdLength := int(binary.BigEndian.Uint32(packet[pointer : pointer+4]))
	pointer += 4
	log.Println("packetId length:", packetIdLength)
	if packetIdLength > 20000000 {
		return
	}
	packetId := string(packet[pointer : pointer+packetIdLength])
	pointer += packetIdLength
	log.Println("packetId:", packetId)
	payload := packet[pointer:]
	log.Println(string(payload))

	if path == "authenticate" {
		var lis *signaler.Listener
		success, _, _ := t.app.Tools().Security().AuthWithSignature(userId, payload, signature)
		if success {
			lis = t.connectListener(userId)
			var pointIds []string
			t.app.ModifyState(true, func(trx trx.ITrx) {
				pIds, err := trx.GetLinksList("memberof::"+userId+"::", -1, -1)
				if err != nil {
					log.Println(err)
					pointIds = []string{}
				} else {
					pointIds = pIds
				}
			})
			for _, pointId := range pointIds {
				t.app.Tools().Signaler().JoinGroup(pointId, userId)
			}
			t.writeResponse(packetId, 0, packetmodel.BuildErrorJson("authenticated"), false)
			t.app.Tools().Signaler().ListenToSingle(lis)
			b, _ := json.Marshal(packetmodel.ResponseSimpleMessage{Message: "old_queue_end"})
			lis.Signal("old_queue_end", b)
		} else {
			t.writeResponse(packetId, 4, packetmodel.BuildErrorJson("authentication failed"), false)
		}
		return
	}
	action := t.app.Actor().FetchAction(path)
	if action == nil {
		t.writeResponse(packetId, 1, packetmodel.BuildErrorJson("action not found"), false)
		return
	}
	var err error
	input, err := action.(iaction.ISecureAction).ParseInput("tcp", payload)
	if err != nil {
		log.Println(err)
		t.writeResponse(packetId, 2, packetmodel.BuildErrorJson(err.Error()), false)
		return
	}
	statusCode, result, err := action.(iaction.ISecureAction).SecurelyAct(userId, packetId, payload, signature, input, strings.Split(t.Conn.RemoteAddr().String(), ":")[0])
	log.Println(result)
	if err != nil {
		httpStatusCode := 3
		if statusCode == -1 {
			httpStatusCode = 4
		}
		t.writeResponse(packetId, httpStatusCode, packetmodel.BuildErrorJson(err.Error()), false)
	}
	t.writeResponse(packetId, 0, result, false)
}

func NewTcp(app core.ICore) *Tcp {
	m := cmap.New[*Socket]()
	return &Tcp{app: app, sockets: &m}
}

File: node/kasper/drivers/network/tcp.cpp
#include "tcp.h"

using json = nlohmann::json;

template <typename T>
void SafeQueue<T>::push(const T &item)
{
	{
		std::lock_guard<std::mutex> lock(mutex_);
		queue_.push(item);
	}
	cond_var_.notify_one();
}

template <typename T>
void SafeQueue<T>::push(T &&item)
{
	{
		std::lock_guard<std::mutex> lock(mutex_);
		queue_.push(std::move(item));
	}
	cond_var_.notify_one();
}

template <typename T>
T SafeQueue<T>::wait_and_pop()
{
	std::unique_lock<std::mutex> lock(mutex_);
	cond_var_.wait(lock, [this]()
				   { return !queue_.empty(); });

	T item = std::move(queue_.front());
	queue_.pop();
	return item;
}

template <typename T>
std::optional<T> SafeQueue<T>::try_pop()
{
	std::lock_guard<std::mutex> lock(mutex_);
	if (queue_.empty())
	{
		return std::nullopt;
	}

	T item = std::move(queue_.front());
	queue_.pop();
	return item;
}

template <typename T>
bool SafeQueue<T>::empty() const
{
	std::lock_guard<std::mutex> lock(mutex_);
	return queue_.empty();
}

template <typename T>
size_t SafeQueue<T>::size() const
{
	std::lock_guard<std::mutex> lock(mutex_);
	return queue_.size();
}

template <typename T>
ValPack SafeQueue<T>::peek() const
{
	std::lock_guard<std::mutex> lock(mutex_);
	return queue_.front();
}

void SocketItem::writeRawUpdate(std::string key, char *updatePack, uint32_t len)
{
	std::cerr << "preparing update..." << std::endl;

	const char *keyBytes = key.c_str();
	auto keyBytesLen = Utils::getInstance().convertIntToData(key.size());

	uint32_t packetSize = 1 + 4 + key.size() + len;
	auto packet = new char[packetSize];
	uint32_t pointer = 1;

	packet[0] = 0x01;

	memcpy(packet + pointer, keyBytesLen, 4);
	pointer += 4;
	delete keyBytesLen;
	memcpy(packet + pointer, keyBytes, key.size());
	pointer += key.size();
	memcpy(packet + pointer, updatePack, len);
	pointer += len;

	std::cerr << "appending to buffer..." << std::endl;

	std::lock_guard<std::mutex> lock(this->lock);
	this->buffer.push({packet, packetSize});
	this->pushBuffer();
}

void SocketItem::writeObjUpdate(std::string key, json updatePack)
{
	std::string data = updatePack.dump();
	this->writeRawUpdate(key, &data[0], data.size());
}

void SocketItem::writeRawResponse(std::string requestId, int resCode, char *response, uint32_t len)
{
	std::cerr << "preparing response..." << std::endl;

	const char *b1 = requestId.c_str();
	char *b1Len = Utils::getInstance().convertIntToData(requestId.size());

	char *b2 = Utils::getInstance().convertIntToData(resCode);

	uint32_t packetSize = 1 + 4 + requestId.size() + 4 + len;
	char *packet = new char[packetSize];

	uint32_t pointer = 1;

	packet[0] = 0x02;

	memcpy(packet + pointer, b1Len, 4);
	pointer += 4;
	delete b1Len;
	memcpy(packet + pointer, b1, requestId.size());
	pointer += requestId.size();
	memcpy(packet + pointer, b2, 4);
	pointer += 4;
	memcpy(packet + pointer, response, len);
	pointer += len;

	std::cerr << "appending to buffer..." << std::endl;

	std::lock_guard<std::mutex> lock(this->lock);
	this->buffer.push({packet, packetSize});
	this->pushBuffer();
}

void SocketItem::writeObjResponse(std::string requestId, int resCode, json response)
{
	std::string data = response.dump();
	this->writeRawResponse(requestId, resCode, &data[0], data.size());
}

void SocketItem::pushBuffer()
{
	if (this->ack)
	{
		if (this->buffer.size() > 0)
		{
			this->ack = false;
			auto data = this->buffer.peek();
			char *packetLen = Utils::getInstance().convertIntToData(data.len);
			auto res = send(this->conn, packetLen, 4, 0);
			delete packetLen;
			if (res == -1)
			{
				this->ack = true;
				std::cerr << "error writing to socket." << std::endl;
			}
			send(this->conn, data.data, data.len, 0);
			if (res == -1)
			{
				this->ack = true;
				std::cerr << "error writing to socket." << std::endl;
			}
		}
	}
}

std::function<void(std::string, std::any, size_t)> SocketItem::connectListener(std::string uid)
{
	auto lis = [this](std::string key, std::any data, size_t len)
	{
		if (len == 0)
		{
			this->writeObjUpdate(key, std::any_cast<json>(data));
		}
		else
		{
			this->writeRawUpdate(key, std::any_cast<char *>(data), len);
		}
	};
	this->ack = true;
	return lis;
}

void SocketItem::processPacket(char *packet, uint32_t len)
{
	if (len == 1 && packet[0] == 0x01)
	{
		std::lock_guard<std::mutex> lock(this->lock);
		this->ack = true;
		if (this->buffer.size() > 0)
		{
			auto top = this->buffer.peek();
			this->buffer.try_pop();
			delete top.data;
			this->pushBuffer();
		}
		return;
	}

	std::cerr << "received packet length: " << len << std::endl;

	std::string signature = "";
	std::string userId = "";
	std::string path = "";
	std::string packetId = "";
	DataPack payload;

	uint32_t pointer = 0;
	char *tempBytes = new char[4];
	memcpy(tempBytes, packet + pointer, 4);
	std::cerr << "kasper " << static_cast<int>(packet[0]) << " " << static_cast<int>(packet[1]) << " " << static_cast<int>(packet[2]) << " " << static_cast<int>(packet[3]) << std::endl;
	uint32_t signatureLength = Utils::getInstance().parseDataAsInt(tempBytes);
	delete tempBytes;
	std::cerr << "signature length: " << signatureLength << std::endl;
	pointer += 4;
	if (signatureLength > 0)
	{
		char *sign = new char[signatureLength];
		memcpy(sign, packet + pointer, signatureLength);
		signature = std::string(sign, signatureLength);
		delete sign;
		pointer += signatureLength;
	}
	std::cerr << "signature: " << signature << std::endl;

	char *tempBytes3 = new char[4];
	memcpy(tempBytes3, packet + pointer, 4);
	uint32_t userIdLength = Utils::getInstance().parseDataAsInt(tempBytes3);
	delete tempBytes3;
	std::cerr << "userId length: " << userIdLength << std::endl;
	pointer += 4;
	if (userIdLength > 0)
	{
		char *tempBytes4 = new char[userIdLength];
		memcpy(tempBytes4, packet + pointer, userIdLength);
		userId = std::string(tempBytes4, userIdLength);
		delete tempBytes4;
		pointer += userIdLength;
	}
	std::cerr << "userId: " << userId << std::endl;

	char *tempBytes5 = new char[4];
	memcpy(tempBytes5, packet + pointer, 4);
	uint32_t pathLength = Utils::getInstance().parseDataAsInt(tempBytes5);
	delete tempBytes5;
	std::cerr << "path length: " << pathLength << std::endl;
	pointer += 4;
	if (pathLength > 0)
	{
		char *tempBytes6 = new char[pathLength];
		memcpy(tempBytes6, packet + pointer, pathLength);
		path = std::string(tempBytes6, pathLength);
		delete tempBytes6;
		pointer += pathLength;
	}
	std::cerr << "path: " << path << std::endl;

	char *tempBytes7 = new char[4];
	memcpy(tempBytes7, packet + pointer, 4);
	uint32_t packetIdLength = Utils::getInstance().parseDataAsInt(tempBytes7);
	delete tempBytes7;
	std::cerr << "packetId length: " << packetIdLength << std::endl;
	pointer += 4;
	if (packetIdLength > 0)
	{
		char *tempBytes8 = new char[packetIdLength];
		memcpy(tempBytes8, packet + pointer, packetIdLength);
		packetId = std::string(tempBytes8, packetIdLength);
		delete tempBytes8;
		pointer += packetIdLength;
	}
	std::cerr << "packetId: " << packetId << std::endl;

	char *payloadRaw = new char[len - pointer];
	memcpy(payloadRaw, packet + pointer, len - pointer);
	payload = DataPack{payloadRaw, len - pointer};
	std::cerr << "payload: " << payload.data << std::endl;

	try
	{
		if (path == "authenticate")
		{
			std::function<void(std::string, std::any, size_t)> lis;
			auto checkRes = this->core->getTools()->getSecurity()->authWithSignature(userId, std::string(payload.data, payload.len), &signature[0]);
			if (checkRes.verified)
			{
				lis = this->connectListener(userId);
				json res;
				res["message"] = "authenticated";
				this->writeObjResponse(packetId, 0, res);
				this->core->getTools()->getSignaler()->listenToSingle(userId, lis);
				json obj;
				lis("old_queue_end", obj, 0);
			}
			else
			{
				json res;
				res["message"] = "authentication failed";
				this->writeObjResponse(packetId, 4, res);
			}
			delete payloadRaw;
			return;
		}
		auto action = this->core->getActor()->findActionAsSecure(path);
		if (action == NULL)
		{
			json res;
			res["message"] = "action not found";
			this->writeObjResponse(packetId, 1, res);
			delete payloadRaw;
			return;
		}
		auto response = action->run(this->core->getIp(), [this](std::function<void(StateTrx *)> fn)
									{ this->core->modifyState(fn); }, core->getTools(), userId, payload, signature);
		if (response.err != "")
		{
			json data;
			data["message"] = response.err;
			this->writeObjResponse(packetId, response.resCode, data);
			delete payloadRaw;
			return;
		}
		this->writeObjResponse(packetId, 0, response.data);
	}
	catch (const std::exception &e)
	{
		std::cerr << "Standard exception caught: " << e.what() << std::endl;
		json data;
		data["message"] = e.what();
		this->writeObjResponse(packetId, 2, data);
	}
	catch (...)
	{
		std::cerr << "Unknown exception caught" << std::endl;
	}
	delete payloadRaw;
}

Tcp::Tcp(ICore *core)
{
	this->core = core;
	this->idCounter = 0;
	this->sockets = {};
}

std::shared_future<void> Tcp::run(int port)
{
	std::cerr << "starting tcp server..." << std::endl;
	return std::async(std::launch::async, [port, this]
					  {
			int serverSocket = socket(AF_INET, SOCK_STREAM, 0);
			sockaddr_in serverAddress;
			serverAddress.sin_family = AF_INET;
			serverAddress.sin_port = htons(port);
			serverAddress.sin_addr.s_addr = INADDR_ANY;
			bind(serverSocket, (struct sockaddr*)&serverAddress,
				 sizeof(serverAddress));
			listen(serverSocket, 5);
			while (true)
			{
				int clientSocket = accept(serverSocket, nullptr, nullptr);
				std::cerr << "new client connected." << std::endl;
				auto id = this->idCounter++;
				std::thread t([this, clientSocket, id]{
					this->handleConnection(id, clientSocket);
					this->sockets.erase(id);
					close(clientSocket);
				});
				t.detach();
	 		} })
		.share();
}

void Tcp::handleConnection(uint64_t connId, int conn)
{
	auto socket = new SocketItem{conn, {}, true, this->core};
	this->sockets.insert({connId, socket});
	char lenBuf[4];
	char buf[1024];
	char nextBuf[2048];
	uint64_t readCount = 0;
	uint64_t oldReadCount = 0;
	bool enough = false;
	bool beginning = true;
	uint32_t length = 0;
	uint32_t readLength = 0;
	uint32_t remainedReadLength = 0;
	char *readData;
	int counter = 0;
	while (true)
	{
		if (!enough)
		{
			readLength = recv(conn, buf, sizeof(buf), 0);
			if (readLength == 0)
			{
				std::cerr << "socket closed" << std::endl;
				return;
			}
			else if (readLength == -1)
			{
				std::cerr << "socket error" << std::endl;
				return;
			}
			oldReadCount = readCount;
			readCount += readLength;
			memcpy(nextBuf + remainedReadLength, buf, readLength);
			remainedReadLength += readLength;
		}

		if (beginning)
		{
			if (readCount >= 4)
			{
				memcpy(lenBuf, nextBuf, 4);
				remainedReadLength -= 4;
				memcpy(nextBuf, nextBuf + 4, remainedReadLength);
				length = Utils::getInstance().parseDataAsInt(lenBuf);
				readData = new char[length];
				memset(readData, 0, length);
				readCount -= 4;
				beginning = false;
				enough = true;
			}
			else
			{
				enough = false;
			}
		}
		else
		{
			if (remainedReadLength == 0)
			{
				enough = false;
			}
			else if (readCount >= length)
			{
				std::cerr << std::endl
						  << "keyhan " << oldReadCount << " " << length << " " << static_cast<int>(nextBuf[0]) << " " << static_cast<int>(nextBuf[1]) << " " << static_cast<int>(nextBuf[2]) << " " << static_cast<int>(nextBuf[3]) << " " << std::endl
						  << std::endl;
				memcpy(readData + oldReadCount, nextBuf, length - oldReadCount);
				memcpy(nextBuf, nextBuf + (readLength - (readCount - length)), readCount - length);
				remainedReadLength = (readCount - length);
				std::cerr << std::endl
						  << "keyhan -- " << static_cast<int>(readData[0]) << " " << static_cast<int>(readData[1]) << " " << static_cast<int>(readData[2]) << " " << static_cast<int>(readData[3]) << " " << std::endl
						  << std::endl;
				std::cerr << "packet received" << std::endl;
				char *packet = new char[length];
				memcpy(packet, readData, length);
				delete readData;
				std::cerr << std::endl
						  << "konstantin -- " << static_cast<int>(packet[0]) << " " << static_cast<int>(packet[1]) << " " << static_cast<int>(packet[2]) << " " << static_cast<int>(packet[3]) << " " << std::endl
						  << std::endl;

				std::thread t([&socket, length](char *packet)
							  {
								   try
								   {
									   socket->processPacket(packet, length);
								   }
								   catch (const std::exception &e)
								   {
									   std::cerr << "Standard exception caught: " << e.what() << std::endl;
								   }
								   catch (...)
								   {
									   std::cerr << "Unknown exception caught" << std::endl;
								   }
								   delete packet; }, packet);
				t.detach();
				readCount -= length;
				enough = true;
				beginning = true;
			}
			else
			{
				memcpy(readData + oldReadCount, nextBuf, readCount - oldReadCount);
				remainedReadLength = 0;
				enough = true;
			}
		}
	}
}
